factor/telegram/classes/factor.php

```diff
@@ -0,0 +1,315 @@
+            $a = new \stdClass();
+            $a->sitename = 'Moodle'; // TODO
+            $a->code = $newcode;
+            $message = get_string('telegram:message', 'factor_telegram', $a);
+            $telegram->send_message($telegramuserid, $message);
+
+        } else if ($record->timecreated + $duration < time()) {
+            // Old code found. Keep id, update fields.
+            $DB->update_record('tool_mfa', array(
+                'id' => $record->id,
+                'secret' => $newcode,
+                'label' => $_SERVER['HTTP_USER_AGENT'],
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ));
+            $instanceid = $record->id;
+            $token = get_config('factor_telegram', 'telegrambottoken');
+            $telegram = new telegram($token);
+            $a = new \stdClass();
+            $a->sitename = 'Moodle'; // TODO
+            $a->code = $newcode;
+            $message = get_string('telegram:message', 'factor_telegram', $a);
+            $telegram->send_message($telegramuserid, $message);
+        }
+    }
+
+    /**
+     * Verifies entered code against stored DB record.
+     *
+     * @return bool
+     */
+    private function check_verification_code($enteredcode) {
+        global $DB, $USER;
+        $duration = get_config('factor_telegram', 'duration');
+
+        // Get instance that isnt parent email type (label check).
+        // This check must exclude the main singleton record, with the label as the email.
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
+        $record = $DB->get_record_sql($sql, array($USER->id, 'telegram'));
+
+        if ($enteredcode == $record->secret) {
+            if ($record->timecreated + $duration > time()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Cleans up email records once MFA passed.
+     *
+     * {@inheritDoc}
+     */
+    public function post_pass_state() {
+        global $DB, $USER;
+        // Delete all telegram records except base record.
+        $selectsql = 'userid = ?
+                  AND factor = ?
+                   AND label NOT LIKE \'telegram:%\'';
+        $DB->delete_records_select('tool_mfa', $selectsql, array($USER->id, 'telegram'));
+
+        // Update factor timeverified.
+        parent::post_pass_state();
+    }
+
+    /**
+     * TOTP Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_factor_form_definition($mform) {
+        $mform->addElement('text', 'telegramuserid', get_string('telegram:telegramuserid', 'factor_telegram'));
```

Comment:
I found this a little confusing:

1) the form asks for an ID which I usually assume is a number. I could not find that number anywhere so I used myusername which appeared to work

2) there was no validation that I own this username. It should send a code to this account and you must validate / verify that code before the telegram record should be created. See the TOTP form for an example of this

Reply:
Ah as it turned out using my username didn't work at all. I had to go digging but eventually found my user id. I think this should accept a username and then make an API call the grab the userid and then store that so it works if people change their name.

Also the dud factor I setup using my username could not be revoked. Nor could the second one I made using my user id:

![image](https://user-images.githubusercontent.com/187449/91239201-39d13c80-e782-11ea-8030-d4dc3423ace1.png)

Lastly the 'label' for the dud was somehow my user agent string which is weird, it should show my username / name.